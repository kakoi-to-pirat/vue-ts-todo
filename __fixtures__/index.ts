import TreeNodeModel from '@/models/TreeNodeModel';
export const StaticUsersRoles: TreeNodeModel[] = [
  {
    id: 5,
    parentId: 0,
    name: 'Management',
    isApprove: true,
  },
  {
    id: 2,
    parentId: 3,
    name: 'Complete Removal',
    isApprove: true,
  },
  {
    id: 4,
    parentId: 5,
    name: 'Modify',
    isApprove: true,
  },
  {
    id: 3,
    parentId: 4,
    name: 'Removal',
    isApprove: false,
  },
  {
    id: 113,
    name: 'Preview',
    isApprove: false,
  },
  {
    id: 112,
    parentId: 4,
    name: 'Edit',
    isApprove: true,
  },
  {
    id: 110,
    parentId: 4,
    name: 'Add',
    isApprove: true,
  },
];
