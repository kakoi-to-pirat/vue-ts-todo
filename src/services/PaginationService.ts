import { Injectable } from '@vue-ioc/core';
import PaginationServiceModel from '@/models/PaginationServiceModel';

@Injectable()
export class PaginationService implements PaginationServiceModel {
  public pageLast(pagesTotal: number, pagesLimit: number): number {
    return pagesTotal / pagesLimit;
  }

  public pagePrev(currentPage: number): number {
    return currentPage - 1 || 1;
  }

  public pageNext(currentPage: number, pageLast: number): number {
    const next = currentPage + 1;
    return next < pageLast ? next : pageLast;
  }
}
