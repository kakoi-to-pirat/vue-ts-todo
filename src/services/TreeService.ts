import { Injectable } from '@vue-ioc/core';
import * as _ from 'lodash';
import TreeServiceModel from '@/models/TreeServiceModel';
import TreeNodeModel from '@/models/TreeNodeModel';
import UsersRoleModel from '@/models/UsersRoleModel';

@Injectable()
export class TreeService implements TreeServiceModel {
  build(data: TreeNodeModel[]): TreeNodeModel[] {
    const tree = data.reduce((acc: TreeNodeModel[], node: TreeNodeModel) => {
      if (this.isChild(node)) {
        return acc;
      }

      const children = this.getNodeChildrenFromData(data, node.id);
      const TreeNode = this.getTreeNode(node, children);

      return [...acc, TreeNode];
    }, []);

    return tree;
  }

  serializeNode(node: TreeNodeModel): string {
    return _.keys(node).reduce((acc: string, prop: string) => {
      const value = _.get(node, prop);
      return acc ? `${acc} ${prop}: ${value}, ` : `${prop}: ${value}, `;
    }, '');
  }

  withoutChildren(node: TreeNodeModel): TreeNodeModel {
    return _.omit(node, ['children']);
  }

  parse(tree: TreeNodeModel[]): TreeNodeModel[] {
    return tree.reduce((acc: TreeNodeModel[], node: TreeNodeModel) => {
      const children = _.get(node, 'children');
      return children
        ? [
            ...acc,
            ...this.parse(children as TreeNodeModel[]),
            this.withoutChildren(node),
          ]
        : [...acc, node];
    }, []);
  }

  update(tree: TreeNodeModel[], node: TreeNodeModel): TreeNodeModel[] {
    if (node.children) {
      node.children = node.children.map(item =>
        this.updateChildApprove(item as UsersRoleModel, node.isApprove),
      );
    }

    return this.updateTree(tree, node);
  }

  isIndeterminate(node: TreeNodeModel): boolean {
    const children = _.get(node, 'children', []);
    return children.some(child => !_.get(child, 'isApprove'));
  }

  private isChild(node: TreeNodeModel): boolean {
    return Boolean(_.get(node, 'parentId'));
  }

  private isNodeChild(child: TreeNodeModel, nodeId: number): boolean {
    const childParentId = _.get(child, 'parentId');
    return childParentId === nodeId;
  }

  private getNodeChildrenFromData(
    data: TreeNodeModel[],
    nodeId: number,
  ): TreeNodeModel[] {
    const children = data.reduce(
      (children: TreeNodeModel[], child: TreeNodeModel) =>
        this.isNodeChild(child, nodeId) ? [...children, child] : children,
      [],
    );

    const grandChildren = children.map((child: TreeNodeModel) => {
      const currentChildren = this.getNodeChildrenFromData(data, child.id);
      return this.getTreeNode(child, currentChildren);
    });

    return grandChildren;
  }

  private getTreeNode(
    node: TreeNodeModel,
    children: TreeNodeModel[],
  ): TreeNodeModel {
    return _.isEmpty(children) ? { ...node } : { ...node, children };
  }

  private updateChildApprove(
    child: UsersRoleModel,
    isApprove: boolean,
  ): TreeNodeModel {
    child.isApprove = isApprove;

    if (child.children) {
      child.children = child.children.map(item =>
        this.updateChildApprove(item as UsersRoleModel, isApprove),
      );
    }

    return child;
  }

  private updateTree(
    tree: TreeNodeModel[],
    node: TreeNodeModel,
  ): TreeNodeModel[] {
    return tree.map(nodeChild => {
      if (nodeChild.id === node.id) {
        return node;
      }
      if (nodeChild.children) {
        nodeChild.children = this.updateTree(
          _.get(nodeChild, 'children') as TreeNodeModel[],
          node,
        );
      }
      return nodeChild;
    });
  }
}
