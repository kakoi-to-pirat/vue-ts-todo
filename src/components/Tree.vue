<template>
  <div class="tree">
    <div class="tree__data">
      <h3>Source:</h3>
      <tree-node :tree="data" />

      <h3>Result:</h3>
      <tree-node :tree="tree" />
    </div>

    <div class="tree__actions">
      <h3>Editable:</h3>
      <ul class="tree__actions-list">
        <tree-node
          :tree="tree"
          :is-editable="true"
          :editable-field="editableField"
          @change-node="onChangeNode"
        />
      </ul>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue } from 'vue-property-decorator';
import { Inject } from '@vue-ioc/core';
import { TreeService } from '@/services/TreeService';
import TreeNodeModel from '@/models/TreeNodeModel';
import TreeNode from '@/components/TreeNode.vue';
@Component({
  components: { TreeNode },
})
export default class Tree extends Vue {
  @Inject()
  public treeService!: TreeService;

  @Prop()
  private data!: TreeNodeModel[];

  @Prop({ default: 'checked' })
  private editableField!: string;

  get tree(): TreeNodeModel[] {
    return this.treeService.build(this.data);
  }

  onChangeNode(node: TreeNodeModel) {
    const updatedTree = this.treeService.update(this.tree, node);
    const parsedTree = this.treeService.parse(updatedTree);
    this.$emit('change', parsedTree);
  }
}
</script>

<style scoped>
h3 {
  margin: 40px 0 0;
}
ul {
  list-style-type: none;
  padding: 0;
}
li {
  display: inline-block;
  margin: 0 10px;
}
a {
  color: #42b983;
}
.tree {
  display: flex;
  justify-content: space-between;
  border: 1px solid #eee;
  padding: 5px;
}
.tree__data {
  width: calc(50% - 2.5px);
  border: 1px solid #eee;
  padding: 0 15px;
  margin-right: 2.5px;
  text-align: left;
}
.tree__actions {
  width: calc(50% - 2.5px);
  border: 1px solid #eee;
  padding: 0 15px;
  margin-left: 2.5px;
  text-align: left;
}
</style>
