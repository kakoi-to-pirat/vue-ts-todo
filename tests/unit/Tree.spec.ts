import { mount, createLocalVue } from '@vue/test-utils';
import Tree from '@/components/Tree.vue';
import { TreeService } from '@/services/TreeService';
import { StaticUsersRoles } from '../../__fixtures__';

const localVue = createLocalVue();

const wrapper = mount(Tree, {
  localVue,
  mocks: { treeService: new TreeService() },
  propsData: { data: StaticUsersRoles, editableField: 'isApprove' },
});

describe('Tree.vue', () => {
  it('set props.data when passed', () => {
    expect(wrapper.isVueInstance()).toBe(true);
    expect(wrapper.props('data')).toEqual(StaticUsersRoles);
  });
  it('shold match the snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
