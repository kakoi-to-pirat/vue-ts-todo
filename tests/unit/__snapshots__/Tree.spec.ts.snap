// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree.vue shold match the snapshot 1`] = `
<div class="tree">
  <div class="tree__data">
    <h3>Source:</h3>
    <ul class="tree-list">
      <li class="tree-list__item"><span>id: 5,  parentId: 0,  name: Management,  isApprove: true, </span>
        <!---->
      </li>
      <li class="tree-list__item"><span>id: 2,  parentId: 3,  name: Complete Removal,  isApprove: true, </span>
        <!---->
      </li>
      <li class="tree-list__item"><span>id: 4,  parentId: 5,  name: Modify,  isApprove: true, </span>
        <!---->
      </li>
      <li class="tree-list__item"><span>id: 3,  parentId: 4,  name: Removal,  isApprove: false, </span>
        <!---->
      </li>
      <li class="tree-list__item"><span>id: 113,  name: Preview,  isApprove: false, </span>
        <!---->
      </li>
      <li class="tree-list__item"><span>id: 112,  parentId: 4,  name: Edit,  isApprove: true, </span>
        <!---->
      </li>
      <li class="tree-list__item"><span>id: 110,  parentId: 4,  name: Add,  isApprove: true, </span>
        <!---->
      </li>
    </ul>
    <h3>Result:</h3>
    <ul class="tree-list">
      <li class="tree-list__item"><span>id: 5,  parentId: 0,  name: Management,  isApprove: true,  children: [object Object], </span>
        <ul class="tree-list">
          <li class="tree-list__item"><span>id: 4,  parentId: 5,  name: Modify,  isApprove: true,  children: [object Object],[object Object],[object Object], </span>
            <ul class="tree-list">
              <li class="tree-list__item"><span>id: 3,  parentId: 4,  name: Removal,  isApprove: false,  children: [object Object], </span>
                <ul class="tree-list">
                  <li class="tree-list__item"><span>id: 2,  parentId: 3,  name: Complete Removal,  isApprove: true, </span>
                    <!---->
                  </li>
                </ul>
              </li>
              <li class="tree-list__item"><span>id: 112,  parentId: 4,  name: Edit,  isApprove: true, </span>
                <!---->
              </li>
              <li class="tree-list__item"><span>id: 110,  parentId: 4,  name: Add,  isApprove: true, </span>
                <!---->
              </li>
            </ul>
          </li>
        </ul>
      </li>
      <li class="tree-list__item"><span>id: 113,  name: Preview,  isApprove: false, </span>
        <!---->
      </li>
    </ul>
  </div>
  <div class="tree__actions">
    <h3>Editable:</h3>
    <ul class="tree__actions-list">
      <ul class="tree-list">
        <li class="tree-list__item"><span><input type="checkbox">
      Management
    </span>
          <ul class="tree-list">
            <li class="tree-list__item"><span><input type="checkbox">
      Modify
    </span>
              <ul class="tree-list">
                <li class="tree-list__item"><span><input type="checkbox">
      Removal
    </span>
                  <ul class="tree-list">
                    <li class="tree-list__item"><span><input type="checkbox">
      Complete Removal
    </span>
                      <!---->
                    </li>
                  </ul>
                </li>
                <li class="tree-list__item"><span><input type="checkbox">
      Edit
    </span>
                  <!---->
                </li>
                <li class="tree-list__item"><span><input type="checkbox">
      Add
    </span>
                  <!---->
                </li>
              </ul>
            </li>
          </ul>
        </li>
        <li class="tree-list__item"><span><input type="checkbox">
      Preview
    </span>
          <!---->
        </li>
      </ul>
    </ul>
  </div>
</div>
`;
